Option for separate endpoints for server and client

### What problem does this feature solve?

When running locally, the Apollo client on the server and in the browser can both use `http://localhost:8000/graphql` (or whatever the port is) as the `httpEndpoint` and everything works fine:

``` js
export default {
  apollo: {
    clientConfigs: {
      default: {
        httpEndpoint: `http://${HOST}:${PORT}/graphql`,
      },
    },
  },
}
```

However, when deployed to production (or any other non-local environment), the server can use `http://localhost:8000/graphql`, but the browser definitely cannot. The simplest solution in my case is for the client value to be just `/graphql`, and the browser will assume the same domain and port as the currently-running page.

Both [`@nuxtjs/axios`][@nuxtjs/axios] and [`@nuxt/http`][@nuxt/http] have separate options for the server (`baseURL`) and the browser (`browserBaseURL`). It seems the Apollo module needs the same to support anyone writing an app where their UI and API are running on the same server. (Unless I missed something. Apollo seems to assume that the API is running on a different server.)

[@nuxtjs/axios]: https://axios.nuxtjs.org/options
[@nuxt/http]: https://http.nuxtjs.org/api/

I tried the following workaround in `nuxt.config.js` of using multiple Apollo clients, but I wasn't able to get it working:

``` js
export default {
  modules: [
    "@nuxtjs/apollo",
  ],
  apollo: {
    clientConfigs: {
      default: {
        httpEndpoint: "localhost:8000/graphql",
      },
      browser: {
        httpEndpoint: "/graphql",
      },
    },
  },
};
```

I tried putting the following in every Vue component to dynamically select the right client based on the environment, but got errors in the browser:

``` html
<script>
  export default {
    apollo: {
      $client: typeof window === "undefined" ? "default" : "browser",
      // ...
    },
  };
</script>
```


### What does the proposed changes look like?

In `nuxt.config.js`, just like for the [`@nuxtjs/axios`][@nuxtjs/axios] and [`@nuxt/http`][@nuxt/http] modules, the Apollo module should support separate endpoint values for server and client:

``` js
export default {
  modules: [
    "@nuxtjs/apollo",
  ],
  apollo: {
    clientConfigs: {
      default: {
        httpEndpoint: "localhost:8000/graphql",
        browserHttpEndpoint: "/graphql",
      },
    },
  },
};
```

[@nuxtjs/axios]: https://axios.nuxtjs.org/options
[@nuxt/http]: https://http.nuxtjs.org/api/


<!--cmty--><!--cmty_prevent_hook-->
<div align="right"><sub><em>This feature request is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/apollo-module/issues/c252">#c252</a>)</em></sub></div>

Indeed, this works as I remembered.

I tried adding the following to my `nuxt.config.js`:

``` js
console.log(`process.server = ${process.server}`);
console.log(`process.client = ${process.client}`);

export default {
  modules: [
    "@nuxtjs/apollo",
  ],
  apollo: {
    clientConfigs: {
      default: {
        httpEndpoint: (() => {
          if (process.server) return "http://localhost:8000/graphql";
          if (process.client) return "/graphql";
          throw new Error("Both process.server and process.client are falsy");
        })(),
      },
    },
  },
};
```

I get the following output when I run the build:

``` text
process.server = undefined
process.client = undefined
/Users/ccalo/Projects/viaticus/nuxt.config.js:1
Error: Both process.server and process.client are falsy
    at require.apollo.clientConfigs.default.httpEndpoint (/Users/ccalo/Projects/viaticus/nuxt.config.js:42:17)
    at Object.<anonymous> (/Users/ccalo/Projects/viaticus/nuxt.config.js:43:11)
    at Generator.next (<anonymous>)
    at Object.<anonymous> (/Users/ccalo/Projects/viaticus/server/nuxt.js:1)
    at Generator.next (<anonymous>)
    at Object.<anonymous> (/Users/ccalo/Projects/viaticus/server/index.js:1)
    at Generator.next (<anonymous>)
    at Module._extensions..js (internal/modules/cjs/loader.js:789:10)
[ERROR] 09:00:43 Error
```

This seems like a design flaw in Nuxt, but there might be good reason why this doesn't work.

--- a/lib/templates/plugin.js
+++ b/lib/templates/plugin.js
@@ -55,6 +55,11 @@ export default (ctx, inject) => {
       if (!<%= key %>ClientConfig.getAuth) {
         <%= key %>ClientConfig.getAuth = <%= key %>GetAuth
       }
+
+      if (process.browser && <%= key %>ClientConfig.browserHttpEndpoint) {
+        <%= key %>ClientConfig.httpEndpoint = <%= key %>ClientConfig.browserHttpEndpoint
+      }
+
       <%= key %>ClientConfig.ssr = !!process.server
       <%= key %>ClientConfig.cache = <%= key %>Cache
       <%= key %>ClientConfig.tokenName = <%= key %>TokenName

--- a/lib/templates/plugin.js
+++ b/lib/templates/plugin.js
@@ -56,7 +56,7 @@ export default (ctx, inject) => {
         <%= key %>ClientConfig.getAuth = <%= key %>GetAuth
       }
 
-      if (process.browser && <%= key %>ClientConfig.browserHttpEndpoint) {
+      if (process.client && <%= key %>ClientConfig.browserHttpEndpoint) {
         <%= key %>ClientConfig.httpEndpoint = <%= key %>ClientConfig.browserHttpEndpoint
       }
 

